waku/common/databases/db_sqlite.nim

```diff
@@ -345,3 +345,141 @@ proc setUserVersion*(database: SqliteDatabase, version: int64): DatabaseResult[v
+      if not isSqlScript(script):
+        return false
+
+      if direction != "" and not script.toLower().endsWith("." & direction & ".sql"):
+        return false
+
+      let scriptVersionRes = getMigrationScriptVersion(script)
+      if scriptVersionRes.isErr():
+        return false
+      
+      let scriptVersion = scriptVersionRes.value
+      return lowVersion < scriptVersion and scriptVersion <= highVersion
+
+  paths.filter(filterPredicate)
+
+proc sortMigrationScripts(paths: seq[string]): seq[string] =
+  ## Sort migration scripts paths alphabetically
+  paths.sorted(system.cmp[string])
+
+proc loadMigrationScripts(paths: seq[string]): Result[seq[string], string] =
+  var loadedScripts = newSeq[string]()
+
+  for script in paths:
+    try:
+      loadedScripts.add(readFile(script))
+    except OSError, IOError:
+      return err("failed to load script '" & script & "': " & getCurrentExceptionMsg())
+
+  ok(loadedScripts)
+
+proc breakIntoStatements(script: string): seq[string] =
+  var statements = newSeq[string]()
+
+  for chunk in script.split(';'):
+    if chunk.strip().isEmptyOrWhitespace(): 
+      continue
+
+    let statement = chunk.strip() & ";"
+    statements.add(statement)
+
+  statements
+
+proc migrate*(db: SqliteDatabase,
+              targetVersion: int64,
+              migrationsScriptsDir: string):
+              Result[void, string] =
+  ## Compares the `user_version` of the sqlite database with the provided `targetVersion`, then
+  ## it runs migration scripts if the `user_version` is outdated. The `migrationScriptsDir` path
+  ## points to the directory holding the migrations scripts once the db is updated, it sets the
+  ## `user_version` to the `tragetVersion`.
+  ##
+  ## NOTE: Down migration it is not currently supported
+  let userVersion = ?db.getUserVersion()
+
+  if userVersion == targetVersion:
+    debug "database schema is up to date", userVersion=userVersion, targetVersion=targetVersion
+    return ok()
+  
+  info "database schema is outdated", userVersion=userVersion, targetVersion=targetVersion
+
+  # Load migration scripts
+  var migrationScriptsPaths = ?listSqlScripts(migrationsScriptsDir)
+  migrationScriptsPaths = filterMigrationScripts(migrationScriptsPaths, lowVersion=userVersion, highVersion=targetVersion, direction="up")
+  migrationScriptsPaths = sortMigrationScripts(migrationScriptsPaths)
+  
+  if migrationScriptsPaths.len <= 0:
+    debug "no scripts to be run"
+    return ok()
+
+  let scripts = ?loadMigrationScripts(migrationScriptsPaths)
+
+  # Run the migration scripts
+  for script in scripts:
+
+    for statement in script.breakIntoStatements():
+      debug "executing migration statement", statement=statement
+
+      let execRes = db.query(statement, NoopRowHandler)
+      if execRes.isErr():
+        error "failed to execute migration statement", statement=statement, error=execRes.error
+        return err("failed to execute migration statement")
```

Comment:
Question: what happen if it fail in the middle of a migration does it need to be rolled back or no?

Reply:
Good point. We are assuming that either all the query works or all the query fails but we are not considering the case where the query partially worked in which case we should make sure the database is rolled back to the initial state. However, I'd consider that optimization for further PRs. I'd like to keep this PR with minimal changes, i.e. only rename files or delete the `if` statement that you suggested :)

Reply:
Btw, I just said "just rename files" but precisely at this point, I'm adding content from "waku/common/sqlite/migrations.nim", which I removed. If you are not happy with that I can restore back the 'migrations.nim'.

Reply:
Totally fine to not consider this edge case for now.